stan/math/rev/core/autodiffstackstorage.hpp

```diff
@@ -7,30 +7,77 @@
+#define STAN_THREADS_DEF
+#endif
+
 /**
- * Provides a thread_local singleton if needed. Read warnings below!
- * For performance reasons the singleton is a global static for the
- * case of no threading which is returned by a function. This design
- * should allow the compiler to apply necessary inlining to get
- * maximal performance. However, this design suffers from "the static
- * init order fiasco"[0].  Anywhere this is used, we must be
- * absolutely positive that it doesn't matter when the singleton will
- * get initialized relative to other static variables.  In exchange,
- * we get a more performant singleton pattern for the non-threading
- * case. In the threading case we use the defacto standard C++11
- * singleton pattern relying on a function wrapping a static local
- * variable. This standard pattern is expected to be well supported
- * by the major compilers (as its standard), but it does incur some
- * performance penalty.  There has been some discussion on this; see
- * [1] and [2] and the discussions those PRs link to as well.
+ * This struct always provides access to the autodiff stack using
+ * the singleton pattern. Read warnings below!
+ *
+ * The singleton <code>instance_</code> is a global static pointer,
+ * which is thread local (TLS) if the STAN_THREADS preprocess variable
+ * is defined.
  *
- * These are thread_local only if the user asks for it with
- * -DSTAN_THREADS. This is primarily because Apple clang compilers
- * before 2016 don't support thread_local and the additional
- * performance cost. We have proposed removing support for those[3],
- * and at that time we should evaluate the performance of a switch to
- * thread_local.  If there is no loss in performance, we can remove
- * this ifdef.
+ * The use of a pointer is motivated by performance reasons for the
+ * threading case. When a TLS is used, initialization with a constant
+ * expression at compile time is required for fast access to the
+ * TLS. As the autodiff storage struct is non-POD, its initialization
+ * is a dynamic expression at compile time. These dynamic expressions
+ * are wrapped, in the TLS case, by a TLS wrapper function which slows
+ * down its access. Using a pointer instead allows to initialize at
+ * compile time to <code>nullptr</code>, which is a compile time
+ * constant. In this case, the compiler avoids the use of a TLS
+ * wrapper function.
+ *
+ * For performance reasons we use the __thread keyword on compilers
+ * which support it. The __thread keyword is a GNU compiler-specific
+ * (gcc, clang, Intel) extension which requires initialization with a
+ * compile time constant expression. The C++11 keyword thread_local
+ * does allow for constant and dynamic initialization of the
+ * TLS. Thus, only the __thread keyword gurantees that constant
+ * initialization and it's implied speedup, is used.
+ *
+ * The initialzation of the AD instance at run-time is handled by the
```

Comment:
[comment only -- no action necessary]
@wds15: this seems like a really weird ownership pattern. I see from the code that this is true, but it's almost like you want to use the RAII pattern (looks like you almost rederived that pattern yourself without getting there completely).

I'm not going to hold this up for that, it just seems like an odd pattern.